### ISSUE TITLE
Config parameter "singleInstance" for modes.window

### ISSUE THREAD

**Issue by JustPilz** [neutralinojs/neutralinojs#901]
Add a `singleInstance` to `modes.window` in `neutralino.config.json`  to prevent opening an application if it is already running.

```
 "modes": {
    "window": {
      "title": "hello-world",
      ...
      "singleInstance": true/false,
      ...
    }
 } 
```
![singleInstance](https://user-images.githubusercontent.com/24465747/169701213-af883a72-37db-4793-851f-8d6a21fcff45.png)

**Comment by pegvin** [comment]
interesting stuff, give us some time will reach out

**Comment by pathange-s** [comment]
Hmm, seems like. But, may i know where you would be using this ? I mean, where the user has to be forced not to open multiple windows ?

**Comment by JustPilz** [comment]
> Hmm, seems like. But, may i know where you would be using this ? I mean, where the user has to be forced not to open multiple windows ?

Any application where multiple instances of app are not required. For example, a game launcher

Maybe this parameter can have several options (as in AHK: https://www.autohotkey.com/docs/commands/_SingleInstance.htm)

**Comment by pathange-s** [comment]
Yes, having several parameters will give flexibility. @shalithasuranga always believes in keeping the project simple and not over-engineered. Let him give his views too.

**Comment by shalithasuranga** [comment]
Hey @pathange-s  this feature will be very beneficial seems ;) maybe we can add this config to the root level so that we can limit app instances from any Neutralinojs mode seems (However, we can override it in window mode too ) :tada:

**Comment by pathange-s** [comment]
Do you mean, giving the user the flexibility to limit app instances to a certain number ?
Great.  I will look into this. You can drop any clues if you have any. Where exactly the code has to be modified. Well, I can figure of the rest (but takes little more time)

**Comment by pathange-s** [comment]
@shalithasuranga seems like, implementation would be different on a different architecture, but yeah, the concept remains the same.
- Firstly let me know, will be going with `singleInstance : true/false` or `maxInstances : 5`.
Well, creating a mutex as suggested in first resource provided above ? I read about it and the main usecase of mutex was : _You can use a mutex object to protect a shared resource from simultaneous access by multiple threads or processes._
Using this might not be necessary i guess. The main usecase itself is different. (Inputs are welcomed, I am new to Windows APIs)

The workaround that I felt would be simple and gg is, by using `PID` or processID. If we want to implement `singleInstance : true/false`, just find the the found datatype is not null. Well, if we need to have `maxInstances : 5` find the number of `PIDs` returned corresponding to the application name.

**Comment by shalithasuranga** [comment]
Hey.. @pathange-s  I think the boolean value is enough for now since it's needed to make single-instance apps -- I am not sure there are any use-cases of the integer value. Electron also do [the same](https://www.electronjs.org/docs/latest/api/app#apprequestsingleinstancelockadditionaldata)

Regarding implementation, seems like `lockf` can be used on Unix. I am not too sure which approach is the best on Win32, maybe we can use some approach that doesn't use GUI elements like windows to support all Neutralinojs modes. Perhaps, there is a way similar to `lockf`.

**Comment by pathange-s** [comment]
Could you check out the last paragraph in [here](https://stackoverflow.com/a/5339606). Seems like something important ?!

**Comment by pegvin** [comment]
No we don't have something like that in the framework, also a [guy shared the implementation of it](https://stackoverflow.com/a/26661272/14516016)

**Comment by JustPilz** [comment]
Any updates?

**Comment by JustPilz** [comment]
up

**Comment by chr15m** [comment]
Seems like a useful feature. I am porting a web game to Windows and this would help a lot.

**Comment by Sledmine** [comment]
up

**Comment by zippie-lab** [comment]
up
